src/main/java/ua/lviv/iot/graph/Graph.java

```diff
@@ -0,0 +1,142 @@
+package ua.lviv.iot.graph;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+/**
+ * An oriented graph build as adjacency list with parameterized vertices.
+ * 
+ * @author Khrystyna Vysotska
+ */
+public class Graph<T> {
+	private int numberOfVertices;
+	private List<T> vertices;
+	private List<List<Integer>> neighborPositions;
+
+	public Graph() {
+		vertices = new LinkedList<>();
+		neighborPositions = new LinkedList<>();
+	}
+
+	/**
+	 * <p>
+	 * Checks if vertex is in graph.
+	 * </p>
+	 *
+	 * @param <T> vertex to check
+	 * @return true if vertex is in graph; false otherwise
+	 */
+	public boolean exists(T vertex) {
+		return vertices.contains(vertex);
+	}
+
+	/**
+	 * Adds new vertex to the graph.
+	 * <p>
+	 * <ul>
+	 * <li>If vertex is already in graph - nothing is done.</li>
+	 * </ul>
+	 * </p>
+	 *
+	 * @param <T> vertex to add
+	 * @return nothing
+	 */
+	public void addVertex(T vertex) {
+		if (!exists(vertex)) {
+			vertices.add(vertex);
+			neighborPositions.add(new LinkedList<Integer>());
+			numberOfVertices++;
+		}
+	}
+
+	/**
+	 * <p>
+	 * Creates an edge between two vertices.
+	 * </p>
+	 *
+	 * @param <T> vertex to draw edge from
+	 * @param <T> vertex to draw edge to
+	 * @throws NoSuchElementException if there is no such vertices in graph
+	 * @return nothing
+	 */
+	public void addEdge(T vertexFrom, T vertexTo) {
+		if (!exists(vertexFrom) || !exists(vertexTo)) {
+			throw new NoSuchElementException();
+		}
+		int vertexFromPosition = getPositionOf(vertexFrom);
+		int vertexToPosition = getPositionOf(vertexTo);
+		neighborPositions.get(vertexFromPosition).add(vertexToPosition);
+	}
+
+	/**
+	 * <p>
+	 * Finds positions of neighbors for this vertex position.
+	 * </p>
+	 *
+	 * @param vertexPosition the position of vertex to get neighbors positions for
+	 * @throws NoSuchElementException    if there is no such vertex in the graph
+	 * @throws IndexOutOfBoundsException if the position is out of range
+	 * @return List<Integer> with positions of neighbors for this vertex position
+	 */
+	public List<Integer> getNeighborPositionsFor(int vertexPosition) {
+		if (exists(vertices.get(vertexPosition))) {
+			return neighborPositions.get(vertexPosition);
+		} else {
+			throw new NoSuchElementException();
+		}
+	}
+
+	/**
+	 * <p>
+	 * Finds vertex by its position.
+	 * </p>
+	 *
+	 * @param vertexPosition the position of vertex to find
+	 * @throws NoSuchElementException    if there is no such vertex in the graph
+	 * @throws IndexOutOfBoundsException if the position is out of range
+	 * @return vertex value for this vertex position
+	 */
+	public T getVertexAt(int vertexPosition) {
```

Comment:
Same thing – very ugly function. You fetch a Vertex by index in order to ... check if it exists (but you just fetched it by index!!!) ... in order to fetch it by index AGAIN! :)

Now, funny story, do you know why with all the checks you do this function is still not prone to errors? Do write a unit test for it and do a test for `getVertexAt(-1)` or `getVertexAt(Integer.MAX_VALUE)` and see what happens